03_Modules/Locations/src/module/admin.ts

```diff
@@ -0,0 +1,172 @@
+  OcpiLocationProps,
+  OcpiLocationRepository,
+  OcpiEvse,
+  OcpiEvseRepository,
+  OcpiConnector,
+  OcpiConnectorRepository,
+  VariableAttributesUtil,
+  OcpiLocationsUtil,
+  InvalidParamException,
+} from '@citrineos/ocpi-base';
+
+@Service()
+export class AdminLocationsService {
+  constructor(
+    private logger: Logger<ILogObj>,
+    private locationMapper: CitrineOcpiLocationMapper,
+    private locationRepository: SequelizeLocationRepository,
+    private ocpiLocationRepository: OcpiLocationRepository,
+    private ocpiEvseRepository: OcpiEvseRepository,
+    private ocpiConnectorRepository: OcpiConnectorRepository,
+    private locationsBroadcaster: LocationsBroadcaster,
+    private variableAttributesUtil: VariableAttributesUtil,
+    private ocpiLocationsUtil: OcpiLocationsUtil,
+  ) {}
+
+  public async createOrUpdateLocation(
+    adminLocationDto: AdminLocationDTO,
+    broadcast: boolean,
+  ): Promise<LocationDTO> {
+    this.logger.debug(
+      adminLocationDto.id
+        ? `Updating Location ${adminLocationDto.id}`
+        : `Creating Location ${adminLocationDto.name}`,
+    );
+
+    const [invalidLocation, message] = AdminLocationDTO.IS_LOCATION_INVALID(adminLocationDto);
+    if (invalidLocation) {
+      throw new InvalidParamException(message ?? 'Invalid location request.');
+    }
+
+    const [ocpiLocation, citrineLocation] =
+      this.mapAdminLocationDtoToEntities(adminLocationDto);
+    const savedCitrineLocation =
+      await this.locationRepository.createOrUpdateLocationWithChargingStations(
+        citrineLocation,
+      );
+    ocpiLocation[OcpiLocationProps.citrineLocationId] = savedCitrineLocation.id;
+    const savedOcpiLocation =
+      await this.ocpiLocationRepository.createOrUpdateOcpiLocation(
+        ocpiLocation,
+      );
+
+    for (const adminEvse of adminLocationDto.evses ?? []) {
+      await this.ocpiEvseRepository.createOrUpdateOcpiEvse(
+        this.mapAdminEvseDtoToOcpiEvse(adminEvse),
+      );
+      for (const adminConnector of adminEvse.connectors ?? []) {
+        await this.ocpiConnectorRepository.createOrUpdateOcpiConnector(
+          this.mapAdminConnectorToOcpiConnector(adminConnector, adminEvse),
+        );
+      }
+    }
+
+    let chargingStationVariableAttributesMap = {};
+
+    if (savedCitrineLocation.chargingPool) {
+      chargingStationVariableAttributesMap =
+        await this.variableAttributesUtil.createChargingStationVariableAttributesMap(
+          savedCitrineLocation.chargingPool.map((station) => station.id),
+        );
+      savedOcpiLocation.ocpiEvses =
+        await this.ocpiLocationsUtil.createOcpiEvsesInfoMap(
+          chargingStationVariableAttributesMap,
+        );
+    }
+
+    const locationDto = this.locationMapper.mapToOcpiLocation(
+      savedCitrineLocation,
+      chargingStationVariableAttributesMap,
+      savedOcpiLocation,
+    );
+
+    if (broadcast) {
+      await this.locationsBroadcaster.broadcastOnLocationCreateOrUpdate(
```

Comment:
Don't we have DB listeners that should emit a broadcast when the DB updates above are made?

Reply:
we don't currently have DB listeners that broadcast automatically bc we want to give some flexibility to the user so that they can choose to broadcast (sometimes you just want to edit something without telling the MSPs), regardless if a database change was made. it's kind of like a "force publish" trigger, but perhaps we can revisit with a better approach 🤔